added necessary fields for event pulling and an event example

ozdoganoguzhan opened the pull request:
**Type Changes:** Added some fields that are required to create a subscription for pulling messages with a filter, and also for pulling messages from that subscription. 

**New Example:** Added an example on how to pull messages periodically from a specific camera. Tested on Dahua.

![Screenshot_10](https://github.com/lumeohq/onvif-rs/assets/30651481/3e565101-c082-4ff9-93af-4cd01b5957b1)

Solves #107

ozdoganoguzhan committed:
added necessary fields for event pulling

edited readme added event example

revision + added SourceType

test fix

DmitrySamoylov commented on onvif/Cargo.toml:
```diff
@@ -36,6 +36,7 @@ xsd-macro-utils = { git = "https://github.com/lumeohq/xsd-parser-rs", rev = "7f3
 xsd-types = { git = "https://github.com/lumeohq/xsd-parser-rs", rev = "7f3d433" }
 yaserde = "0.7.1"
 yaserde_derive = "0.7.1"
+b_2 = {path = "../wsdl_rs/b_2"}
```
It's only used in the example, so let's move it to `[dev-dependencies]`.

ozdoganoguzhan replied:
done

DmitrySamoylov commented on wsdl_rs/b_2/src/lib.rs:
```diff
@@ -131,15 +137,45 @@ pub struct NotificationMessageHolderType {
 
+#[derive(Default, PartialEq, Debug, YaSerialize, YaDeserialize)]
+#[yaserde(prefix = "tt", namespace = "tt: http://www.onvif.org/ver10/schema")]
+
```
Unneeded newline.

ozdoganoguzhan replied:
removed

DmitrySamoylov commented on wsdl_rs/b_2/src/lib.rs:
```diff
@@ -36,7 +39,10 @@ impl Validate for TopicExpressionType {}
-pub struct FilterType {}
+pub struct FilterType {
+    #[yaserde(prefix = "wsnt", rename = "TopicExpression")]
+    pub topic_expression: TopicExpressionType,
```
In WSDL, it's a sequence of `any` types:

```
  <xsd:complexType name="FilterType">
    <xsd:sequence>
      <xsd:any minOccurs="0" maxOccurs="unbounded"/>
    </xsd:sequence>
  </xsd:complexType>
```

so it shouldn't be a required field. For a lack of better handling of `any` types, let's make the field optional to not break other use-cases:

```
pub topic_expression: Option<TopicExpressionType>,
```

ozdoganoguzhan replied:
done

DmitrySamoylov commented on wsdl_rs/b_2/src/lib.rs:
```diff
@@ -131,15 +137,45 @@ pub struct NotificationMessageHolderType {
 pub mod notification_message_holder_type {
     use super::*;
 
+    #[derive(Default, PartialEq, Debug, YaSerialize, YaDeserialize)]
+    #[yaserde(prefix = "tt", namespace = "tt: http://www.onvif.org/ver10/schema")]
+    pub struct DataType {
+        #[yaserde(prefix = "tt", rename = "SimpleItem")]
+        pub simple_item: SimpleItemType,
+    }
+
+    #[derive(Default, PartialEq, Debug, YaSerialize, YaDeserialize)]
+    #[yaserde(prefix = "tt", namespace = "tt: http://www.onvif.org/ver10/schema")]
+    pub struct MessageTypeInner {
+        #[yaserde(prefix = "tt", rename = "Data")]
+        pub data: DataType,
+    }
+
```
Do you have an example XML from the camera? If yes, can you also add a test [here](https://github.com/lumeohq/onvif-rs/blob/a49b43d4713ee5ff5ffb7bab9105d6793c6c1085/schema/src/tests.rs)?

ozdoganoguzhan replied:
Added 2 tests: one for pulling messages one for creating subscription and getting the address from that subscription.

Thank you for great feedback!

DmitrySamoylov commented:
Can you please rebase on `main` to see if CI is green?

ozdoganoguzhan committed:
Merge branch 'lumeohq:main' into main

ozdoganoguzhan committed:
!fixup test xml

asuper0 commented:
hi, how is it going?

ozdoganoguzhan committed:
fix tests

ozdoganoguzhan committed:
Merge branch 'main' into main

ozdoganoguzhan commented:
The problem was in the test XML. I removed the envelope declarations and just left the responses and put the namespaces in them. Also moved the related tests to the end of tests file, to prevent conflict with the more recent commits to test file.

DmitrySamoylov commented:
> error: useless use of `vec!`

This was fixed in `main`. Please, rebase again.

ozdoganoguzhan committed:
Merge branch 'lumeohq:main' into main

sqrtsanta commented:
Sorry for chipping in, also very interested in subscriptions & pull_messages. Really appreciate your work folks!

DmitrySamoylov commented:
Not sure if you receive notifications on failing CI jobs, but clippy is not happy now with some of the changes.

ozdoganoguzhan committed:
fix clippy errors and warnings in examples/event.rs, add event example to readme

ozdoganoguzhan committed:
Merge branch 'lumeohq:main' into main

ozdoganoguzhan commented:
everything should be ok now, thanks for your patience 😄
